### ISSUE TITLE
New field to add OP_RETURN output

### ISSUE THREAD

**Issue by pedromvpg** [sparrowwallet/sparrow#97]
Is there a way to add new field on the "Send" screen to add OP_RETURN output to a transaction?

**Comment by craigraw** [comment]
I'd like to support this, but there isn't space on the Send screen for the area indicated - it's used by the transaction diagram. 

This is how Electrum does it - too hacky?

![](https://pbs.twimg.com/media/EybzbvmWQAASWBk?format=png&name=small)

**Comment by pedromvpg** [comment]
Could it be a control per output, to pre-populate the "Pay to" field?
Pay "MAX" would have to move to the left but would work because it would couple it with the Amount field.

![image](https://user-images.githubusercontent.com/807505/114776747-d01d9700-9d40-11eb-86d6-573334f81f0b.png)

**Comment by 6102bitcoin** [comment]
Outstanding Action: Identify way of adding OpReturn functionality without compromising UI
Proposed Priority: Low

**Comment by inpharmaticist** [comment]
Bump, glad I searched for this. I agree this would be cool, but can't justify any kind of urgency beyond that :laughing:

**Comment by Surenic** [comment]
It is urgent because I'd like to create a pre-signed transaction with an OP_Return for my son's upcoming birth 🙈😂

I won't have time to create but publish a transaction then.

**Comment by somoza** [comment]
+1 to this request. The previous link is not working anymore.

**Comment by dadofsambonzuki** [comment]
+1

**Comment by windsok** [comment]
+1 

Also would be good to make OP_RETURN outputs visible in the transaction viewer

**Comment by streamofstars** [comment]
> Fwiw I still haven't heard a good use for this feature request.

Good use of a feature that allows adding an OP_RETURN message or good use of a specific implementation described in one of the posts in this thread?

**Comment by craigraw** [comment]
> Good use of a feature that allows adding an OP_RETURN message or good use of a specific implementation described in one of the posts in this thread?

A good reason for adding arbitrary OP_RETURN messages. Obviously the BIP47 implementation adds OP_RETURNs already for a specific purpose.

**Comment by streamofstars** [comment]
> A good reason

my reason was to "label" a transaction permanently in the blockchain but I see how this is not necessarily the desirable use case in general, and adding a box for OP_Return would probably lead to unnecessary overuse. On the other hand, it's people's money, so...

It would be nice to have this option, I don't mind if it requires an extra step of checking a box somewhere in the settings or if it is implemented similar way to your electrum example from the second post.

**Comment by Cortexico** [comment]
OP_RETURN outputs are now commonly used around BRC-20 ecosystem and different protocols. I would advise to implement this as an option feature, and make it hidden on the UI natively.

**Comment by MoellerDi** [comment]
> Fwiw I still haven't heard a good use for this feature request.

The lack of a function for adding an OP_RETURN message to a transaction could force users to use dubious websites that offer this service for a fee (and potentially collect our data), or to use other tools simply because they need to prove, for example, that a document related to the transaction existed at that time.

I'm thinking of the use case of collecting SATs for my grandson, granddaughter, godchild, child, etc. In my region, it's important to prove the time of such a donation. A simple transaction on the blockchain isn't enough, as the authorities could argue that the donation only occurred when the wallet is handed over in 5, 10, or 15 years. Who knows what value Bitcoin will have then, but it could result in a huge tax burden.

To avoid this, one could simply handwrite a donation certificate, take a photo, or scan it as a PDF, generate a hash from it, and insert this hash into the OP_RETURN message of the first transaction (or all transactions, depending on the specific case) to the giveaway wallet. This way, even in 5, 10, or 15 years (assuming the photo/PDF still exists), it would be possible to prove that the transaction was intended as a donation. This could save a lot of taxes in the future.

Think of the OP_RETURN message as a kind of "purpose", "intended use", "purpose of use", or "reason for payment". It's like what we're all used to with bank account transactions, but with a big warning that the transaction content is visible to everyone on the blockchain. It could even be an advanced feature and hidden by default.

Ultimately, it helps maintain my privacy, as I don't need a third party or other tools for it; it's just me, Sparrow, and my full node.
I would be very happy to have such a feature in Sparrow.

**Comment by inpharmaticist** [comment]
> > Fwiw I still haven't heard a good use for this feature request.
> 
> The lack of a function for adding an OP_RETURN message to a transaction could force users to use dubious websites that offer this service for a fee (and potentially collect our data), or to use other tools simply because they need to prove, for example, that a document related to the transaction existed at that time.
> 
> I'm thinking of the use case of collecting SATs for my grandson, granddaughter, godchild, child, etc. In my region, it's important to prove the time of such a donation. A simple transaction on the blockchain isn't enough, as the authorities could argue that the donation only occurred when the wallet is handed over in 5, 10, or 15 years. Who knows what value Bitcoin will have then, but it could result in a huge tax burden.
> 
> To avoid this, one could simply handwrite a donation certificate, take a photo, or scan it as a PDF, generate a hash from it, and insert this hash into the OP_RETURN message of the first transaction (or all transactions, depending on the specific case) to the giveaway wallet. This way, even in 5, 10, or 15 years (assuming the photo/PDF still exists), it would be possible to prove that the transaction was intended as a donation. This could save a lot of taxes in the future.
> 
> Think of the OP_RETURN message as a kind of "purpose", "intended use", "purpose of use", or "reason for payment". It's like what we're all used to with bank account transactions, but with a big warning that the transaction content is visible to everyone on the blockchain. It could even be an advanced feature and hidden by default.
> 
> Ultimately, it helps maintain my privacy, as I don't need a third party or other tools for it; it's just me, Sparrow, and my full node. I would be very happy to have such a feature in Sparrow.

For the use case you mention, [Open Timestamps](https://opentimestamps.org) would suffice.

**Comment by MoellerDi** [comment]
> For the use case you mention, [Open Timestamps](https://opentimestamps.org/) would suffice.

First of all, thank you for pointing out that there are other tools/projects for proving the existence of a document at a specific point in time. My post wasn't meant to call that into question.

I'll quote myself:
> ... could force users to use dubious websites that offer this service for a fee (and potentially collect our data), or to use other tools ...

The downside to using third-party providers or other tools is that you have to trust someone else to do the work for you.
You have to trust:
- that they did it correctly
- that they aren't collecting your data
- that the third-party provider will be available in 15+ years
- that the toolset will be maintained for the next 15+ years
- ...

I'll quote myself again:
> ... I don't need a third party or other tools ...

If I have the ability to insert the SHA256 hash of a document into the transaction, thereby linking that transaction to a specific document, there's no need for a third party or specific toolset to be available 15+ years from now. The person in the future only needs the image/PDF, a way to calculate the SHA256 hash from it, and access to the blockchain to look up the transaction. It doesn't require Sparrow or its own full node, nor does it require a third party or toolset (except SHA256, but I assume it's available, because Bitcoin would have a problem otherwise).

Don't get me wrong, I agree with you, there are other ways to do this. No question. But that wasn't the point of my post. I just wanted to share my opinion on a `good use for this feature`. It's not always about spam.

**Comment by bayfitt** [comment]
### Reputation System

OP_RETURN is a feature of the Bitcoin blockchain. Its use is up to the user. There are infinite possibilities for what can be done with it. Check out my GitHub repo op_return_encoder, which is what I intend to use it for. This system provides a foundation for recording reputation data in a transparent, immutable way, but its application depends entirely on the user’s needs.

**Comment by MarkJelic** [comment]
This issue has been open since April 2021... Over 4 years... And we are still at "identifying a location"??

<img width="716" height="290" alt="Image" src="https://github.com/user-attachments/assets/614d65cd-7ab7-47d2-a63f-c62dd9e4a188" />

Are we for real? Here is a screenshot of Sparrow on Mac. I have squished it down to as small as it can go. I don't know about you, but there is PLENTY of space for a field like the OP suggested. Pretending like there isn't is disingenuous.

<img width="655" height="707" alt="Image" src="https://github.com/user-attachments/assets/ee24efe0-c5af-4aa2-98bd-2bdf3d618187" />

As to all the arguments or why it is needed... Does the protocol allow it? Then it should be there.  Anyone putting anything in there will be paying for it. That field doesn't need to be one line, either. It should be a re-sizeable to allow multi-line input.

And with the release of Core v30 and OP_Return being upped to 100k, this too should be fully supported, with the option of selecting a file to attach into the OP_Return field. Of course it should have checks to make sure the file is less than 100k bytes, but otherwise it should allow any file type whatsoever. JPEG, PDF, MP3, MP4. All of it.

Not supporting this proposal means you just want to censor other valid Use-Cases that could bring So Much Utility. And the miners will get paid, right?  This is the direction that our beloved Core Developers want to take us. Are you against the All Knowing Core Developers? If so, go run something else!  /s

**Comment by EricGrill** [comment]
I'm interested in implementing this feature. I'll add an optional OP_RETURN field to the Send screen that allows users to embed arbitrary data (up to 80 bytes) in their transactions.

My approach will be to add a collapsible/optional section for the OP_RETURN data input, with appropriate validation for the 80-byte limit. The OP_RETURN output will have 0 value as required by the Bitcoin protocol.

I'll submit a PR once the implementation is ready.

**Comment by craigraw** [comment]
@EricGrill You are welcome to submit a PR, but fair warning - until I am convinced there is a good reason to add arbitrary OP_RETURNs with Sparrow I'm not merging it.

**Comment by MarkJelic** [comment]
> [@EricGrill](https://github.com/EricGrill) You are welcome to submit a PR, but fair warning - until I am convinced there is a good reason to add arbitrary OP_RETURNs with Sparrow I'm not merging it.

Are we to take it that you are not a supporter of the Core move to open OP-Return to 100k?

I think Eric's proposal to add a field to allow 80 bytes is fair and I think should be added in. The main reason being that it allows a note for the transaction, a reminder, a reference... Just like you can on a check. I really don't think there is anything wrong with such a small amount of data to be included, and correct me if I'm wrong, that does add to the fees paid to the miner? I'm OK with that.

I don't think anyone can argue that Sparrow is the premier wallet that is recommended by most users of Bitcoin. Why not include the ability to use the network as designed? (No, I don't support 100k OP_Returns... My note above was sarcastic.)

**Comment by craigraw** [comment]
> Are we to take it that you are not a supporter of the Core move to open OP-Return to 100k?

That's immaterial to this discussion.

> The main reason being that it allows a note for the transaction, a reminder, a reference... Just like you can on a check.

I specified the [BIP 329](https://github.com/bitcoin/bips/blob/master/bip-0329.mediawiki) wallet labels standard for exactly this purpose, which already has substantial and growing adoption across wallets. Storing label information on a blockchain is not a good idea for privacy reasons.

**Comment by bayfitt** [comment]
The reason to implement unlimited op-return sizes is in order to include as many users as possible in sparrow. If you don't include it, then you are introducing the possibility of forking the software (and losing dev money) over this issue. Those who don't care about op-return won't use it, and those who do care about it will be satisfied.

**Comment by MarkJelic** [comment]
> Storing label information on a blockchain is not a good idea for privacy reasons.

Maybe not... But is it up to you to decide if the user should or should not put text in there that is private or not? If a user is silly enough to put their home address in the OP-Return, should we be stopping them or sit back and shake our heads?  To quote yourself, the user's privacy in the way they use this field is "immaterial to this discussion".

There is a facility in the Core implementation of Bitcoin to allow the user to embed information into the OP-Return field. Even Knots allows this. Yet Sparrow, the most recommended wallet in this space, is not allowing users to use this field? Come on, you have to have a better reason than "I worry about user's privacy"!

And although I do NOT support spam on the chain... Fact is that **it is possible**, but only if you pay the Taproot Wizards some fee or such? And then instead of adding it into OP-Return, they spray the UTXO set with the junk which cannot be pruned. Why not take the financial rewards away from companies offering this spam services and just give it directly to the users for free? And don't forget, "The miners will get paid" which "Strengthens the network".

**Comment by nbxl21** [comment]
Multiple people in my local meetup have asked me how to add OP_RETURN outputs to their transactions. All of them use Sparrow and don't understand why it's not included. My recent issue #1911 was the result of me trying to explain to non-technical people how to create a PSBT with Core first and then sign with their usual setup. It works, but it's not easy for the users I'm used to helping, and it often requires debugging individual setups.

The use cases were time-stamping (I'm aware of OpenTimestamps) and printing public messages (just small easter eggs for the recipients).

In the end, not having a simple Sparrow integration means:
- I need to dedicate time to helping people (I'm happy to do it)
- People need to install additional software or mess with the terminal
- Too much information is leaked to me such as TXIDs, recipients, or details about individual setups. This point is the reason for sending this message.

**Comment by inpharmaticist** [comment]
@nbxl21 are you familiar with [op_return bot](https://opreturnbot.com)? Seems like a better option for nontechnical people.

**Comment by SATutxo** [comment]
Hello, good morning. I’ve just become an uncle and I’d like to create a new mini Bitcoiner by sending a first transaction with an OP_RETURN as a memory of this beautiful day, so that in the future, when he’s older, he can look back and see that first transaction (the genesis of his wallet).

Could someone please explain to me, in a simple and easy way, how to create this transaction in Sparrow from my wallet, where I send him a small UTXO and additionally include an OP_RETURN with a message? Is it better to send the message in hexadecimal?

Thank you very much. I’m reading you.

**Comment by nbxl21** [comment]
> Could someone please explain to me, in a simple and easy way, how to create this transaction in Sparrow from my wallet, where I send him a small UTXO and additionally include an OP_RETURN with a message? Is it better to send the message in hexadecimal?

If you have a node and have access to `bitcoin-cli` the easiest way for me was to create the transaction with bitcoin core and sign it with sparrow like that:

```bitcoin-cli createpsbt '[{"txid": "66c4c0e1219429b81762d6d03574cd73b9176c69a9af6a8a92ad4418480a5240", "vout": 0}]' '[{"tb1q5ma0txehapykm2fvd462wyjvek3vj2kg3y4rmy": 0.00009}, {"data": "48656c6c6f20576f726c6421"}]'```

This command will create a transaction spending one UTXO, send 0.00009 to "tb1q5ma0txehapykm2fvd462wyjvek3vj2kg3y4rmy" and add a OP_RETURN output with "Hello World!" (encoded in hex). You will have to compute the fee manually and add a change output if needed.

Then just import it in sparrow and it will handle it well. If the funds are spent from a hardware wallet you will probably need to upgrade sparrow to the latest version.

**Comment by saxtron3000** [comment]
Heading to a wedding tomorrow and would like to gift sats with a message. Gonna have to use Electrum for the first time ever, it looks like.

**Comment by MarkJelic** [comment]
> 
> If you have a node and have access to `bitcoin-cli` the easiest way for me was to create the transaction with bitcoin core and sign it with sparrow like that:
> 

Really should not have to jump through these kind of hoops when a field could be so easily added to the Sparrow UI.

**Comment by tigernero79** [comment]
Is there any hope that the possibility of inserting an op_return in sparrow wallet will be implemented?

**Comment by MarkJelic** [comment]
I think it is likely. Not by the original developer (OD), who seems to have some issue with allowing the use of standard functionality supported by Bitcoin.  But I get the feeling it would not be hard to implement. Could possibly even Vibe code it or use Grok or such. Just need to fork this repo and do the changes. There have been a lot of forks... Could even have been implemented by one of them, already. May be worth searching through them.

Will OD ever merge those changes into base Sparrow code? I doubt it, again because he has some issue that he has not clearly or truthfully outlined. But I could see a fork of Sparrow that supports Op_Return could quickly garner a large portion of the user base.  What to call it... How about "Sparrow OP" ?

**Comment by d00g** [comment]
Some of this nagging seems so ungrateful.  Craig does an amazing job building and maintaining sparrow.  If you want OP that bad then solve the UI problem and fork/build it yourself.  Maybe if you do a nice enough job and create a PR then he might merge it.  Until then, enjoy the beautiful free software that Craig has provided.

**Comment by MarkJelic** [comment]
Not ungrateful at all. And I'm not even asking Craig to do the development. But he HAS indicated that he has some issue with it and would be unlikely to merge a PR with it included. His reasoning has not been clear. None of what I said above is in any way intended to be ungrateful or "Nagging"... It pretty simply only states facts.

**Comment by tigernero79** [comment]
@d00g 

Sparrow is a masterpiece of software; no one disputes that. The idea of integrating a function recognized as a de facto standard was mentioned. That's all, if it's not possible, so be it. There are alternatives, of course, like Electrum Wallet. But anyone who's attached to a particular software would only want to use it. Different functions aren't being asked for. What's being asked is to integrate a function that has always been accepted by the Bitcoin protocol.

**Comment by d00g** [comment]
The use (or misuse) of OP is much more philosophical and I'm sure you know this.  If you don't like his choice then fork and build what you like.

**Comment by tigernero79** [comment]
You're fixated on the issue of forking. A fork isn't necessary; there are alternatives like the Electrum wallet. The question is bringing that alternative to Sparrow Wallet, which isn't difficult to understand. We know you can fork or use other wallets, but the point is to implement a function that isn't new but is established by the protocol, even on Sparrow Wallet. I hope that's clear now.

**Comment by d00g** [comment]
Fork/don't fork.  Use electrum or anything else. Or use labels within sparrow the way Craig intended it.  Either way, please stop nagging an open source developer to add features he doesn't want to add.

**Comment by tigernero79** [comment]
Nag? A question is asked, stop, you're making it long. It's up to him to say it, not you. That said, let's end the matter.

**Comment by etkaar** [comment]
Looking for the same feature.

(But I'll accept it if the developer intentionally decides to not implement it since Sparrow is a masterpiece of software)

**Comment by flagge-gethub** [comment]
I'd definitely like to see this added as well....sort of like a memo field on a check....limited to 83 bytes of course! :)

**Comment by djpnewton** [comment]
Dont limit to 83 bytes. Larger values could be useful for stopping replay on other chains for example
